File: assembler.py
import random
import collections
import matplotlib.pyplot as plt


def generate_sequence(length):
    '''
    Generate and return a random sequence of the given length.
    '''
    bases = ['A', 'T', 'C', 'G']
    sequence = ''.join(random.choice(bases) for i in range(length))
    return sequence


def hamming_distance(seq1, seq2):
    '''
    Return the Hamming distance between two sequences.
    If seq1 and seq2 have different length, returns None.
    '''
    if len(seq1) != len(seq2):
        return None
    else:
        distance = sum(1 for base1, base2 in zip(seq1, seq2) if base1 != base2)
        return distance


def get_kmers(sequence, k):
    '''
    Return all substrings of length k from the given sequence.
    '''
    kmers = [sequence[i : i + k] for i in range(len(sequence) - k + 1)]
    return kmers


def count_kmers(kmers):
    '''
    Transform list of kmers to a kmer_counts dictionary. 
    Unique kmers are the keys, and their number of appearances in the kmers list are the values. 
    '''
    kmer_counts = {}
    for kmer in kmers:
        kmer_counts[kmer] = 0
    for kmer in kmers:
        kmer_counts[kmer] += 1
    return kmer_counts


def get_unique_counts(kmer_counts):
    '''
    Return all unique kmer counts (and the number of kmers with this count) present in the kmer_counts dictionary.
    '''
    unique_counts = {}
    for kmer, count in kmer_counts.items():
        if not (count in unique_counts):
            unique_counts[count] = 1
        else:
            unique_counts[count] += 1
    unique_counts_list = []
    for count, num_kmers in unique_counts.items():
        unique_counts_list.append((count, num_kmers))
    return unique_counts_list


def list_kmers(kmer_counts):
    '''
    Transform kmer_counts dictionary to a list of kmers.
    '''
    kmers = []
    for kmer, count in kmer_counts.items():
        for i in range(count):
            kmers.append(kmer)
    return kmers


def mutate(kmers, num_mutations):
    '''
    Mutate a random base of a random kmer, with the number of mutations given by num_mutations.
    '''
    bases = ['A', 'T', 'C', 'G']
    # Do not mutate original kmer list, but create a new list instead
    kmers = list(kmers)
    for i in range(num_mutations):
        kmer_index = random.choice(range(len(kmers)))
        base_index = random.choice(range(len(kmers[kmer_index])))
        kmers[kmer_index] = kmers[kmer_index][0 : base_index] + random.choice(bases) + kmers[kmer_index][base_index + 1:]
    return kmers


def estimate_coverage(kmer_counts, threshold=0):
    '''
    Given a kmer_counts dictionary and a threshold for noise, estimate the number of sequences present.
    Any kmer that appears less than the threshold amount of times, will be assumed to be an error/noise and will be discarded.
    '''
    count_values = {}
    for kmer, count in kmer_counts.items():    
        count_values[count] = 0
    for kmer, count in kmer_counts.items():    
        count_values[count] += 1    
    max_value = 0
    coverage = 0
    for count, value in count_values.items():
        if value > max_value and count > threshold:
            max_value = value
            coverage = count
    return coverage


def reduce_kmers(kmer_counts, coverage):
    '''
    Given a kmer_counts dictionary and a coverage, return a scaled-down set of kmers that represent one instance of the sequence.
    '''
    kmers = []
    for kmer, count in kmer_counts.items():
        remainder = count % coverage
        if remainder < (coverage - remainder):
            count = int((count - remainder) / coverage)
        else:
            count = int((count + (coverage - remainder)) / coverage)
        for i in range(count):
            kmers.append(kmer)
    return kmers


def de_bruijn_graph(kmers):
    '''
    Return a De Bruijn graph from the given kmers.
    '''
    graph = {}
    for kmer in kmers:
        k = len(kmer)
        k_m1, k_m2 = get_kmers(kmer, k - 1)
        # Each node records the number of incoming nodes and a deque of outgoing nodes
        graph[k_m1] = {'in' : 0, 'out' : collections.deque([])}
        graph[k_m2] = {'in' : 0, 'out' : collections.deque([])}
    for kmer in kmers:
        k = len(kmer)
        k_m1, k_m2 = get_kmers(kmer, k - 1)
        graph[k_m2]['in'] += 1
        graph[k_m1]['out'].append(k_m2)
    return graph


def eulerian_path(graph):
    '''
    Use Hierholzer's algorithm to find and return an Eulerian path in the given graph.
    If an Eulerian path does not exist, returns None.
    '''
    path = collections.deque()
    edge_count = 0
    start_count = 0
    end_count = 0
    equal_count = 0
    current_node = ""
    # Find valid start node, and determine path existence
    for node, edge_nodes in graph.items():
        edge_count += len(edge_nodes['out'])
        if len(edge_nodes['out']) - edge_nodes['in'] == 1:
            current_node = node
            start_count += 1
        elif edge_nodes['in'] - len(edge_nodes['out']) == 1:
            end_count += 1
        elif edge_nodes['in'] == len(edge_nodes['out']):
            equal_count += 1
    # Return None if an Eulerian path does not exist
    correct_degree_count = start_count + end_count + equal_count
    if start_count > 1 or end_count > 1 or correct_degree_count != len(graph):
        return None
    # If no valid start node was determined, we can choose any node with non-zero degree
    # We look for the first node with an outgoing edge and assign this as the start node
    if not current_node:
        for node, edge_nodes in graph.items():
            if len(edge_nodes['out']) > 0:
                current_node = node
                break
    history = collections.deque()
    while len(path) != edge_count + 1:
        if len(graph[current_node]['out']) == 0:
            if len(history) != 0:
                next_node = history[-1]
                history.pop()
            path.appendleft(current_node)
        else:
            next_node = graph[current_node]['out'][0]
            graph[current_node]['out'].popleft()
            history.append(current_node)
        current_node = next_node
    return path


def join(path):
    '''
    Join together the given Eulerian path to form a sequence.
    '''
    path = list(path)
    sequence = path[0]
    for i, node in enumerate(path[1:], start=1):
        k = len(path[i - 1])
        sequence += node[k - 1:]
    return sequence


def kmer_count_distribution(kmers, threshold=0, plot=False):
    '''
    Given a set of kmers, return a list describing the distribution of kmer count values (the number of times a kmer appears) in the set.
    The list can also be optionally plotted by setting plot=True.
    '''
    kmer_counts = count_kmers(kmers)
    kmer_counts_dist = sorted([(count, num_kmers) for (count, num_kmers) in get_unique_counts(kmer_counts) if count >= threshold], key=lambda x: x[0])
    if plot:
        plt.scatter(*zip(*kmer_counts_dist))
        plt.xlabel("kmer count")
        plt.ylabel("# kmers")
        plt.show()
    return kmer_counts_dist


def estimate_coverage_reduce_kmers(kmers, threshold=0, describe=False):
    '''
    Estimate the coverage of a set of kmers on an unknown sequence.
    Then, reduce the set of kmers down to a coverage value of 1 on the unknown kmers.
    '''
    if describe:
        print('Finding kmer counts...', end=' ', flush=True)
        kmer_counts = count_kmers(kmers)
        print('done.')
        print('Estimating coverage...', end=' ', flush=True)
        coverage_estimate = estimate_coverage(kmer_counts, threshold=threshold)
        print('done.')
        print('Scaling down kmer coverage...', end=' ', flush=True)
        kmers = reduce_kmers(kmer_counts, coverage_estimate)
        print('done.')
    else:
        kmer_counts = count_kmers(kmers)
        kmers = reduce_kmers(kmer_counts, estimate_coverage(kmer_counts, threshold=threshold))
    return kmers


def assemble(kmers, describe=False):
    '''
    Assemble a sequence from the given kmers.
    '''
    if describe:
        print('Generating De Bruijn graph...', end=' ', flush=True)
        graph = de_bruijn_graph(kmers)
        print('done.')
        print('Finding Eulerian path...', end=' ', flush=True)
        path = eulerian_path(graph)
        print('done.')
        print('Joining sequence...', end=' ', flush=True)
        sequence = join(path)
        print('done.')
    else:
        sequence = join(eulerian_path(de_bruijn_graph(kmers)))
    return sequence

File: assembler.cpp
#include <iostream>
#include <string>
#include <random>
#include <unordered_map>
#include <deque>
#include <chrono>
#include <limits>
#include <algorithm>


typedef std::unordered_map<std::string, std::pair<std::deque<std::string>, std::deque<std::string>>> directedGraph;
typedef std::chrono::_V2::system_clock::time_point timePoint;


float duration(timePoint start, timePoint end)
{
    return float(std::chrono::duration_cast<std::chrono::microseconds>(end - start).count()) / 1000000;
}


std::string genSequence(int length, bool time = false) 
{
    /* 
    Generate and return a random sequence (of the given length) of bases.
    */
    timePoint start = std::chrono::high_resolution_clock::now();
    std::string bases = "ATCG";
    std::random_device rd;
    std::mt19937 rng(rd());
    std::uniform_int_distribution<int> uniformDist(0, 3);
    std::string sequence = "";
    sequence.reserve(length);
    int randomInteger;
    for (int i = 0; i < length; i++) 
    {
        randomInteger = uniformDist(rng);
        sequence += bases[randomInteger];
    }
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "genSequence: " << duration(start, end) << " seconds" << std::endl; 
    return sequence;
}


int hammingDistance(std::string &seq1, std::string &seq2, bool time = false)
{
    /*
    Return the Hamming distance between two sequences.
    */
    timePoint start = std::chrono::high_resolution_clock::now();
    int distance;
    if (seq1.length() != seq2.length()) 
        distance = -1;
    else 
    {
        distance = 0;
        int seqLength = seq1.length();
        for (int i = 0; i < seqLength; i++) 
        {
            if (seq1[i] != seq2[i]) 
                distance += 1;
        }
    }
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "hammingDistance: " << duration(start, end) << " seconds" << std::endl; 
    return distance;
}


std::vector<std::string> genKmers(std::string &sequence, int k, bool time = false)
{
    /*
    Return all substrings of length k from the given sequence.
    */
    timePoint start = std::chrono::high_resolution_clock::now();
    std::vector<std::string> kmers;
    if ((sequence.length() == 0) || (k == 0) || (sequence.length() < k))
        kmers.push_back("");
    else
    {
        int numKmers = sequence.length() - k + 1;
        kmers.reserve(numKmers);
        std::string kmer = sequence.substr(0, k);
        kmers.push_back(kmer);
        int kmerEndIndex = k;
        for (int i = 1; i < numKmers; i++)
        {
            kmer = kmers.back();
            kmer.erase(0, 1);
            kmer += sequence[kmerEndIndex];
            kmers.push_back(kmer);
            kmerEndIndex += 1;
        }
    }
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "genKmers: " << duration(start, end) << " seconds" << std::endl; 
    return kmers;
}


void shuffleKmers(std::vector<std::string> &kmers, bool time = false)
{
    /*
    Shuffle a set of kmers.
    */
    timePoint start = std::chrono::high_resolution_clock::now();
    auto rng = std::default_random_engine {};
    std::shuffle(std::begin(kmers), std::end(kmers), rng);
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "shuffleKmers: " << duration(start, end) << " seconds" << std::endl; 
}


directedGraph deBruijnGraph(std::vector<std::string> &kmers, bool time = false)
{
    /*
    Return a De Bruijn graph from the given kmers.
    */
    timePoint start = std::chrono::high_resolution_clock::now();
    directedGraph graph;
    if (kmers.size() == 0)
    {
        graph[""].first.push_back("");
        graph[""].second.push_back("");
    }
    else
    {
        int kMinus1 = kmers[0].length() - 1;
        int numKmers = kmers.size();
        for (int i = 0; i < numKmers; i++)
        {
            std::string kMinus1Left = kmers[i].substr(0, kMinus1);
            std::string kMinus1Right = kmers[i].substr(1, kMinus1);
            graph[kMinus1Right].first.push_back(kMinus1Left);
            graph[kMinus1Left].second.push_back(kMinus1Right);
        }
    }
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "deBruijnGraph: " << duration(start, end) << " seconds" << std::endl; 
    return graph;
}


std::deque<std::string> eulerianPath(directedGraph &graph, bool time = false)
{
    /*
    Use Hierholzer's algorithm to find and return an Eulerian path in the given graph.
    If an Eulerian path does not exist, returns an empty path.
    */
    timePoint start = std::chrono::high_resolution_clock::now();
    int edgeCount = 0;
    int startCount = 0;
    int endCount = 0;
    int nonZeroEqualCount = 0;
    std::string currentNode = "";
    std::deque<std::string> path;
    std::unordered_map<std::string, int> outIndexes;
    for (auto &node : graph)
    {
        auto &nodeVal = node.first;
        int inNodesSize = node.second.first.size();
        int outNodesSize = node.second.second.size();
        edgeCount += outNodesSize;
        if (outNodesSize - inNodesSize == 1)
        {
            if (currentNode == "")
                currentNode = nodeVal;
            startCount += 1;
        }
        else if (inNodesSize - outNodesSize == 1)
            endCount += 1;
        else if ((inNodesSize == outNodesSize) && inNodesSize != 0)
            nonZeroEqualCount += 1;
        outIndexes[nodeVal] = 0;
    }
    if ((startCount > 1) || (endCount > 1) || (startCount + endCount + nonZeroEqualCount != graph.size()))
        path.push_back("");
    else
    {
        if (currentNode == "")
        {
            for (auto &node : graph)
            {
                if (node.second.second.size() > 0)
                {
                    currentNode = node.first;
                    break;
                }
            }
        }
        std::deque<std::string> history;
        std::string nextNode;
        int edgeCountPlusOne = edgeCount + 1;
        while (path.size() != edgeCountPlusOne)
        {
            auto &currentOutNodes = graph[currentNode].second;
            auto &currentOutIndex = outIndexes[currentNode];
            if (currentOutIndex == currentOutNodes.size())
            {
                if (history.size() != 0)
                {
                    nextNode = history.back();
                    history.pop_back();
                }
                path.push_front(currentNode);
            }
            else
            {
                nextNode = currentOutNodes[currentOutIndex];
                currentOutIndex += 1;
                history.push_back(currentNode);
            }
            currentNode = nextNode;
        }
    }
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "eulerianPath: " << duration(start, end) << " seconds" << std::endl; 
    return path;
}


std::string join(std::deque<std::string> &path, bool time = false)
{
    /*
    Join together the given Eulerian path to form a sequence.
    */
    timePoint start = std::chrono::high_resolution_clock::now();
    int pathSize = path.size();
    std::string sequence;
    if (pathSize == 0)
        sequence = "";
    else
    {
        sequence = path[0];
        for (int i = 1; i < pathSize; i++)
            sequence += path[i].back();
    }
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "join: " << duration(start, end) << " seconds" << std::endl; 
    return sequence;
}


// Work in progress
int estimateCoverage(std::vector<std::string> &kmers, int threshold, bool time = false)
{
    /*
    Estimate the number of sequences present in a set of kmers.
    Any kmer that appears less than the threshold amount of times in the set will be assumed to be noise.
    */
    timePoint start = std::chrono::high_resolution_clock::now();
    int coverage;
    if (kmers.size() == 0)
        coverage = -1;
    else
    {
        std::unordered_map<std::string, int> kmerFrequencies;
        for (auto &kmer : kmers)
        {
            if (kmerFrequencies.find(kmer) == kmerFrequencies.end())
                kmerFrequencies[kmer] = 1;
            else
                kmerFrequencies[kmer] += 1;
        }
        std::unordered_map<int, int> kmerFrequencyCounts;
        for (auto &pair : kmerFrequencies)
        {
            int &frequency = pair.second;
            if (kmerFrequencyCounts.find(frequency) == kmerFrequencyCounts.end())
                kmerFrequencyCounts[frequency] = 1;
            else
                kmerFrequencyCounts[frequency] += 1;
        }
        coverage = std::numeric_limits<int>::min();
        for (auto &pair : kmerFrequencyCounts)
        {
            auto &frequency = pair.first;
            auto &count = pair.second; 
            if ((count > coverage) && (frequency > threshold))
                coverage = frequency;
        }
    }
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "estimateCoverage: " << duration(start, end) << " seconds" << std::endl; 
    return coverage;
}


// Work in progress
std::vector<std::string> normaliseKmers(std::vector<std::string> &kmers, bool time = false)
{
    timePoint start = std::chrono::high_resolution_clock::now();
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "normaliseKmers: " << duration(start, end) << " seconds" << std::endl;
    return kmers;
}


// Work in progress
std::string assemble(std::vector<std::string> &reads, bool time = false)
{
    /*
    Assemble a sequence from the given reads.
    */
    timePoint start = std::chrono::high_resolution_clock::now();
    int numReads = reads.size();
    if (numReads == 0)
        return "";
    std::vector<std::string> kmers;
    int k = std::numeric_limits<int>::max();
    int readLength;
    for (int i = 0; i < numReads; i++)
    {
        readLength = reads[i].length();
        if (readLength < k)
            k = readLength - 1;
    }
    for (int i = 0; i < numReads; i++)
    {
        std::vector<std::string> readKmers = genKmers(reads[i], k);
        int readKmersSize = readKmers.size();
        for (int j = 0; j < readKmersSize; j++)
            kmers.push_back(readKmers[j]);
    }
    int coverage = estimateCoverage(kmers, 0);
    if (coverage > 1)
        kmers = normaliseKmers(kmers);
    auto graph = deBruijnGraph(kmers);
    auto path = eulerianPath(graph);
    auto sequence = join(path);
    timePoint end = std::chrono::high_resolution_clock::now();
    if (time == true)
        std::cout << "assemble: " << duration(start, end) << " seconds" << std::endl; 
    return sequence;
}


int main()
{
    timePoint start = std::chrono::high_resolution_clock::now();
    std::string seq = genSequence(1000000, true);
    std::vector<std::string> kmers = genKmers(seq, 50, true);
    // kmers.reserve(2 * kmers.size()); // Don't understand why
    // kmers.insert(kmers.end(), kmers.begin(), kmers.end());
    shuffleKmers(kmers);   
    int coverage = estimateCoverage(kmers, 0, true);
    std::cout << "coverage: " << coverage << std::endl;
    // Sequence assembly
    auto graph = deBruijnGraph(kmers, true);
    auto path = eulerianPath(graph, true);
    auto assembledSeq = join(path, true);
    auto dist = hammingDistance(seq, assembledSeq, true);
    std::cout << "dist: " << dist << std::endl;
    timePoint end = std::chrono::high_resolution_clock::now();
    std::cout << "main: " << duration(start, end) << " seconds" << std::endl;
}